OpenRA.Mods.Cnc/Activities/LeapAttack.cs

```diff
@@ -0,0 +1,112 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Cnc.Activities;
+using OpenRA.Mods.Cnc.Traits;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Activities
+{
+	public class LeapAttack : Activity
+	{
+		readonly Target target;
+		readonly bool allowMovement;
+		readonly AttackLeapInfo info;
+
+		readonly Mobile mobile;
+		readonly AttackLeap attack;
+
+		ConditionManager conditionManager;
+
+		public LeapAttack(Actor self, Target target, bool allowMovement, AttackLeapInfo info)
+		{
+			this.target = target;
+			this.info = info;
+			this.allowMovement = allowMovement;
+			mobile = allowMovement ? self.TraitOrDefault<Mobile>() : null;
+			attack = self.Trait<AttackLeap>();
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+		}
+
+		Activity MakeLeapActivity(Actor self)
+		{
+			var origin = self.World.Map.CenterOfSubCell(self.Location, mobile.FromSubCell);
+			var targetMobile = target.Actor.TraitOrDefault<Mobile>();
+			var targetSubcell = targetMobile != null ? targetMobile.FromSubCell : SubCell.Any;
+			var destination = self.World.Map.CenterOfSubCell(target.Actor.Location, targetSubcell);
+			var length = Math.Max((origin - destination).Length / info.Speed.Length, 1);
+
+			var leapToken = ConditionManager.InvalidConditionToken;
+
+			return ActivityUtils.SequenceActivities(
```

Comment:
Please use child activities and OnFirstRun instead of SequenceActivities.  Refer to the Transform activity for an example.